.github/workflows/main.yml

```diff
@@ -186,3 +196,167 @@ jobs:
+    runs-on: ubuntu-20.04
+    steps:
+    - name: Checkout Code
+      uses: actions/checkout@v2
+      with:
+        submodules: 'recursive'
+
+    - name: Install Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+
+    - name: Install Tools
+      run: >
+        sudo apt update || true
+
+        sudo apt install -y -qq
+        python3-docutils
+        python3-pip
+        git
+
+        pip install
+        meson==0.56.2
+        ninja
+        zstandard
+
+    - name: Pull cached emsdk
+      id: cache-emsdk
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.EM_CACHE_FOLDER }}
+        key: ${{ env.EM_VERSION }}-${{ runner.os }}
+
+    - name: Install emsdk
+      if: steps.cache-emsdk.outputs.cache-hit != 'true'
+      run: |
+        rm -rf emsdk/
+        git clone https://github.com/emscripten-core/emsdk.git
+        emsdk/emsdk install ${{ env.EM_VERSION }}
+        emsdk/emsdk activate ${{ env.EM_VERSION }}
+
+    - name: Verify emsdk
+      run: |
+        source emsdk/emsdk_env.sh
+        emcc -v
+        tee misc/ci/emscripten-ephemeral-ci.ini <<EOF >/dev/null
+        [constants]
+        toolchain = '$(pwd)/emsdk/upstream/emscripten/'
+        EOF
+
+    - name: Configure
+      run: >
+        source emsdk/emsdk_env.sh
+
+        meson setup build/
+        --cross-file misc/ci/emscripten-ephemeral-ci.ini
+        --cross-file misc/ci/emscripten-build-test-ci.ini
+        --prefix=$(pwd)/build-test
+
+        meson configure
+        -Dbuild.cpp_std=gnu++14
+        build/
+
+    - name: Build
+      run: |
+        source emsdk/emsdk_env.sh
+        # TODO: SPIRV-Tools has a race condition where it sometimes fails, fix that
+        meson compile tar -C build/ --verbose || true
+        meson compile tar -C build/ --verbose
+
+    - name: Upload Log
+      uses: actions/upload-artifact@v2
+      with:
+        name: taisei_emscripten_stable_build_log
+        path: build/meson-logs/meson-log.txt
+        if-no-files-found: warn
+
+  emscripten-test-build-latest:
+    name: "Emscripten Test (emsdk: experimental) (WebGL)"
+    if: "!contains(github.event.head_commit.message, '[skip ci]')"
+    runs-on: ubuntu-20.04
+    continue-on-error: true
+    steps:
+    - name: Checkout Code
+      uses: actions/checkout@v2
+      with:
+        submodules: 'recursive'
+
+    - name: Install Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+
+    - name: Install Tools
+      run: >
+        sudo apt update || true
+
+        sudo apt install -y -qq
+        python3-docutils
+        python3-pip
+        git
+
+        pip install
+        meson==0.56.2
+        ninja
+        zstandard
+
+    - name: Pull cached emsdk
+      id: cache-emsdk
+      uses: actions/cache@v2
+      with:
+        path: ${{ env.EM_CACHE_FOLDER }}
+        key: ${{ env.EM_VERSION_EXPERIMENTAL }}-${{ runner.os }}
```

Comment:
How is this going to work after emscripten updates? Won't we have a stale cache entry for `latest-whateverOS` with the previous version? Is the cache invalidated after some time?

Reply:
It's still useful to pull the cache, but the command below makes sure to fetch and install the latest version anyway, in case it's been updated. It speeds things up when a new version hasn't dropped.

Reply:
Ok, so it'll install the latest version. But what happens on subsequent builds? Does it pull the old version from the cache and install the new one over it again? Or does it somehow eventually update/invalidate the cache entry?

Reply:
The `latest` in `./emsdk install latest` is actually a tag they change in `main`, as seen here: https://github.com/emscripten-core/emsdk/blob/32fd1fafac571965ce88f778dac1072ed0d7e061/emscripten-releases-tags.txt#L2

If `latest` doesn't change when the step below this one `reset`s against `origin/main`, `install` and `activate` basically do nothing. If `latest` *has* changed, then it'll install (and cache) that new version. It still speeds up the process if the `latest` tag hasn't changed.

You can test it out by doing something like:
```
git checkout 2.0.20
./emsdk install latest
git checkout 2.0.21
./emsdk install latest
```

You'll see that it actually pulls new versions each time. This simulates what would happen if `main` were to change as well.

Reply:
> If `latest` has changed, then it'll install **(and cache)** that new version.

The only thing that's not clear to me is the part in bold. Are we *sure* it will actually cache the new version in this case? Does it *always* update the cached entry, even after a cache hit? That's the only way this setup can work properly, so we need to make sure of that. Though in this case it acts less like a cache and more like a mutable persistent filesystem. Which is a strange thing to have in a controlled CI environment, so it's reasonable to expect that it does not work like that.

Reply:
Unfortunately, you're correct. Good catch.

I had thought that if the cache directory had changed, it would update it with the final cleanup action, but apparently cache hits are immutable. I seriously wonder what GitHub's design philosophy was with their CI, it feels so undercooked.

Reply:
I think it's not necessarily a bad design choice. Having a persistent read-write filesystem equivalent in a supposedly reproducible build setup doesn't make a lot of sense (but then neither does having direct internet access, so i guess it's a moot point).

Anyway, the real design flaw is that it's not possible to compute the cache key dynamically in a previous step. Or is it? I don't actually know. But if it is possible, that is what we actually want: figure out what the latest version is, make a cache key out of it, try to get it from the cache, if missed then install it.

If that's not possible, then I don't think having a test build for the latest version is a good idea. I'm not comfortable with reinstalling emscripten from scratch on every push. It's reasonable to just say we only support one emscripten version — the one we test against — since they break it so often, and just bump it manually occasionally. This is kind of what I've been doing already anyway, except no one knew what emscripten version we're actually targeting.

Reply:
Alright, what I've done in the latest commit is explicitly pull the latest tag out of `emsdk` using `git describe`. It then uses that tag to fetch that version from the cache, and if it isn't found, it `install`/`activate`s. Also, if `EM_VERSION` == `latest`, then the other steps in the job get skipped entirely.